dds_web/api/project.py

```diff
@@ -196,20 +149,172 @@ def post(self):
+                deadline = project.current_deadline
+            else:
+                project.released = current_time
+
+        # Create row in ProjectStatuses
+        return models.ProjectStatuses(
+            status="Available", date_created=current_time, deadline=deadline
+        )
+
+    def retract_project(self, project: models.Project, current_time: datetime.datetime):
+        """Retract project: Make status In Progress.
+
+        Only possible from Available.
+        """
+        # Check if valid status transition
+        self.check_transition_possible(
+            current_status=project.current_status, new_status="In Progress"
+        )
+
+        return models.ProjectStatuses(status="In Progress", date_created=current_time)
+
+    def expire_project(
+        self, project: models.Project, current_time: datetime.datetime, deadline_in: int
+    ) -> models.ProjectStatuses:
+        """Expire project: Make status Expired.
+
+        Only possible Available.
+        """
+        # Check if valid status transition
+        self.check_transition_possible(current_status=project.current_status, new_status="Expired")
+
+        if deadline_in > 30:
+            raise DDSArgumentError(
+                message="The deadline needs to be less than (or equal to) 30 days."
+            )
+
+        deadline = dds_web.utils.current_time(to_midnight=True) + datetime.timedelta(
+            days=deadline_in
+        )
+        return models.ProjectStatuses(
+            status="Expired", date_created=current_time, deadline=deadline
+        )
+
+    def delete_project(self, project: models.Project, current_time: datetime.datetime):
+        """Delete project: Make status Deleted.
+
+        Only possible from In Progress.
+        """
+        # Check if valid status transition
+        self.check_transition_possible(current_status=project.current_status, new_status="Deleted")
+
+        # Can only be Deleted if never made Available
+        if project.has_been_available:
+            raise DDSArgumentError(
+                "Project cannot be deleted if it has ever been made available, abort it instead."
+            )
+        project.is_active = False
+
+        # Deletes files (also commits session in the function - possibly refactor later)
+        RemoveContents().delete_project_contents(project=project)
+        self.rm_project_user_keys(project=project)
+
+        # Delete metadata from project row
+        self.delete_project_info(proj=project)
+        delete_message = (
+            f"\nAll files in project '{project.public_id}' deleted and project info cleared"
+        )
+
+        return models.ProjectStatuses(status="Deleted", date_created=current_time), delete_message
+
+    def archive_project(
+        self, project: models.Project, current_time: datetime.datetime, aborted: bool = False
+    ):
+        """Archive project: Make status Archived.
+
+        Only possible from Available and Expired.
```

Comment:
From a quick glance, it looks like you're allowing it to be archived from "In Progress" as well. Not sure how it's supposed to be?

Reply:
I can change this. You can only abort. Which is checked below in
```
if not aborted and project.current_status == "In Progress":
```

Reply:
Hmm I'll check. It's suppose to only be possible to abort.

Reply:
I mean you can Archive, Archived has a `is_aborted` column.

Reply:
So it's actually aborting that's allowed from In Progress

Reply:
But yes, I think I need to change somehting in the code below

Reply:
Changed.